### PR TITLE
feat: derive side effects from package json

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**
### Description
1.  derive side effects from package json
2. the overhead look unavoidable, I will try to optimize it when I have time.

### CODE REVIEW

File: crates/rolldown/src/module_loader/module_loader.rs [hunk 1]
```diff
@@ -24,13 +27,19 @@ use crate::{SharedOptions, SharedResolver};
 
 pub struct IntermediateNormalModules {
   pub modules: IndexVec<NormalModuleId, Option<NormalModule>>,
+  pub module_side_effects: IndexVec<NormalModuleId, Option<Arc<PackageJson>>>,
```
Comment (hyf0) [inline crates/rolldown/src/module_loader/module_loader.rs]: We probably don't need this to get package.json.

Comment (IWANABETHATGUY) [reply crates/rolldown/src/module_loader/module_loader.rs]: OK, is there another place we already store the related `PackageJson`, I can't find the related logic in our codebase.

Comment (hyf0) [reply crates/rolldown/src/module_loader/module_loader.rs]: The code is placed in resolver. Let me create a PR to show it. I'm almost done it.

Comment (hyf0) [reply crates/rolldown/src/module_loader/module_loader.rs]: https://github.com/rolldown/rolldown/pull/1170

File: crates/rolldown/src/stages/link_stage/tree_shaking.rs [hunk 2]
```diff
@@ -10,21 +10,25 @@ struct Context<'a> {
   symbols: &'a Symbols,
   is_included_vec: &'a mut IndexVec<NormalModuleId, IndexVec<StmtInfoId, bool>>,
   is_module_included_vec: &'a mut IndexVec<NormalModuleId, bool>,
+  has_export_used: &'a mut IndexVec<NormalModuleId, bool>,
```
Comment (hyf0) [inline crates/rolldown/src/stages/link_stage/tree_shaking.rs]: What does `has_export_used` field do?

**Review by hyf0 — DISMISSED** [review]

**Review by hyf0 — APPROVED** [review]

**Pull request merged**